### PR TITLE
Aws secrets manager

### PR DESCRIPTION
**Pull request by zodraz**
**What this PR does / why we need it**:

Adds support for Healtchecks on AWS Secrets Manager.

**Which issue(s) this PR fixes**:

Please reference the issue this PR will close: #1125

**Special notes for your reviewer**:

None

**Does this PR introduce a user-facing change?**:

No since it is a brand new HealthCheck  project

Please make sure you've completed the relevant tasks for this PR, out of the following list:

### CODE REVIEW

**Comment by sungam3r** [comment]
Please resolve conflicts.

File: src/HealthChecks.Aws.SecretsManager/HealthChecks.Aws.SecretsManager.csproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetFrameworkVersion);$(NetStandardVersion)</TargetFrameworks>
+    <PackageTags>HealthCheck;AWS;Secrets Manager;</PackageTags>
+    <Description>HealthChecks.Aws.SecretsManager is the health check package for Secrets Manager.</Description>
+    <Version>$(HealthCheckAWSSecretsManager)</Version>
+    <Nullable>enable</Nullable>
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/HealthChecks.Aws.SecretsManager.csproj:8]:
```suggestion
```

NRT annotations are enabled for all projects, see `Directory.Build.props`

**Comment by zodraz** [comment]
Bad conflict merge resolution .Will look over again later. Thanks.

File: AspNetCore.Diagnostics.HealthChecks.sln [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 17
-VisualStudioVersion = 17.0.31903.59
+# Visual Studio Version 16
```
Comment (sungam3r) [inline AspNetCore.Diagnostics.HealthChecks.sln:3]: Could you revert these lines?

File: src/HealthChecks.Aws.SecretsManager/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# AWS Secrets Manager Health Check
+
+This health check verifies the ability to communicate with Amazon Secrets Manager and the existence of some secrets. For more information about AWS Secrets Mananger check the [AWS Secrets Manager Site](https://aws.amazon.com/secrets-manager/)
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/README.md:3]:
```suggestion
This health check verifies the ability to communicate with Amazon Secrets Manager and the existence of some secrets. For more information about AWS Secrets Manager check the [AWS Secrets Manager Site](https://aws.amazon.com/secrets-manager/)
```

File: src/HealthChecks.Aws.SecretsManager/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+- `name`: The health check name. Default if not specified is `aws secrets manager`.
+- `failureStatus`: The `HealthStatus` that should be reported when the health check fails. Default is `HealthStatus.Unhealthy`.
+- `tags`: A list of tags that can be used to filter sets of health checks.
+- `timeout`: A `System.TimeSpan` representing the timeout of the check.
+
+### Basic
+
+### Check existence of a secret and load credentials from the application's default configuration
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/README.md:18]:
```suggestion
```csharp
```

File: src/HealthChecks.Aws.SecretsManager/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        .AddSecretsManager(options =>
+        {
+            options.AddSecret("secretname");
+        });
+}
+```
+
+### Check existence of a secret and directly pass credentials
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/README.md:32]:
```suggestion
```csharp
```

File: src/HealthChecks.Aws.SecretsManager/README.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            options.AddSecret("secretname");
+            options.Credentials = new BasicAWSCredentials("access-key", "secret-key");
+        });
+}
+```
+
+### Check existence of a secret and specify region endpoint
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/README.md:47]:
```suggestion
```csharp
```

File: src/HealthChecks.Aws.SecretsManager/README.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            options.AddSecret("secretname");
+            options.RegionEndpoint = RegionEndpoint.EUCentral1;
+        });
+}
+```
+
+### Check existence of a secret and specify credentials with region endpoint
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/README.md:62]:
```suggestion
```csharp
```

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
+using Amazon.SecretsManager;
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs:1]: Please use file scoped namespaces for all new files.

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+            return client;
+        }
+
+        private async Task<string> GetSecretAsync(IAmazonSecretsManager client, string secretName, CancellationToken cancellationToken)
+        {
+            var request = new GetSecretValueRequest
+            {
+                SecretId = secretName,
+                VersionStage = "AWSCURRENT" // VersionStage defaults to AWSCURRENT if unspecified.
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs:68]: Candidate for new property in `SecretsManagerOptions`?

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+            var response = await client.GetSecretValueAsync(request, cancellationToken);
+
+            // Decrypts secret using the associated KMS key.
+            // Depending on whether the secret is a string or binary, one of these fields will be populated.
+            if (response.SecretString == null)
+            {
+                var memoryStream = response.SecretBinary;
+                var reader = new StreamReader(memoryStream);
+                return System.Text.Encoding.UTF8.GetString(Convert.FromBase64String(await reader.ReadToEndAsync()));
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs:79]: Result of `GetSecretAsync` method is not used. Here you allocate 3 potentially big objects in the heap:
1. `string` from `reader.ReadToEndAsync()`
2. `byte[]` from `Convert.FromBase64String`
3. `string` from `System.Text.Encoding.UTF8.GetString`

Consider change design to only check "good" result without allocate all this stuff.

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+using Amazon.Runtime;
+
+namespace HealthChecks.Aws.SecretsManager
+{
+    public class SecretsManagerOptions
+    {
+        public AWSCredentials? Credentials { get; set; }
+
+        public RegionEndpoint? RegionEndpoint { get; set; }
+
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs:11]:
```suggestion
```

File: .github/workflows/healthchecks_aws_secretsmanager_cd.yml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    - name: Restore dependencies
+      run: dotnet restore ./src/HealthChecks.Aws.SecretsManager/HealthChecks.Aws.SecretsManager.csproj
+    - name: Build
+      run: dotnet build --no-restore ./src/HealthChecks.Aws.SecretsManager/HealthChecks.Aws.SecretsManager.csproj
+    - name: Test
+      run: dotnet test ./test/HealthChecks.Aws.SecretsManager.Tests/HealthChecks.Aws.SecretsManager.Tests.csproj --verbosity normal
+    - name: dotnet pack
+      run: dotnet pack ./src/HealthChecks.Aws.SecretsManager/HealthChecks.Aws.SecretsManager.csproj -c $BUILD_CONFIG --include-source --include-symbols -o ./artifacts
+    - name: setup nuget
+      uses: NuGet/setup-nuget@v1.0.5
```
Comment (sungam3r) [inline .github/workflows/healthchecks_aws_secretsmanager_cd.yml:28]: see #1062

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs (hunk 9 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs:68]: Don't think it is necessary...

File: .github/workflows/healthchecks_aws_secretsmanager_cd.yml (hunk 12 above)
Comment (zodraz) [reply .github/workflows/healthchecks_aws_secretsmanager_cd.yml]: Fixed

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs (hunk 10 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs]: Fixed. Changed to just get the result, will throw if not valid one, so not needed the result actually.

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs (hunk 8 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs:1]: Fixed

File: src/HealthChecks.Aws.SecretsManager/README.md (hunk 7 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/README.md]: Fixed

File: src/HealthChecks.Aws.SecretsManager/README.md (hunk 6 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/README.md]: Fixed

File: src/HealthChecks.Aws.SecretsManager/README.md (hunk 5 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/README.md]: Fixed

File: src/HealthChecks.Aws.SecretsManager/README.md (hunk 4 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/README.md]: Fixed

File: src/HealthChecks.Aws.SecretsManager/README.md (hunk 3 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/README.md]: Fixed

File: AspNetCore.Diagnostics.HealthChecks.sln (hunk 2 above)
Comment (zodraz) [reply AspNetCore.Diagnostics.HealthChecks.sln:3]: Fixed

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs (hunk 9 above)
Comment (sungam3r) [reply src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs]: ok

File: test/HealthChecks.Aws.SecretsManager.Tests/DependencyInjection/RegistrationTests.cs [hunk 13]
```diff
@@ -0,0 +1,135 @@
+using Amazon;
+using Amazon.Runtime;
+using FluentAssertions;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace HealthChecks.Aws.SecretsManager.Tests.DependencyInjection
```
Comment (sungam3r) [inline test/HealthChecks.Aws.SecretsManager.Tests/DependencyInjection/RegistrationTests.cs:9]: file-scoped namespaces in tests oo

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    internal HashSet<string> _secrets = new HashSet<string>();
+
+    internal IEnumerable<string> Secrets => _secrets;
+
+    /// <summary>
+    /// Add an AWS Secrets Manager secret to be checked
+    /// </summary>
+    /// <param name="secretName">The secret to be checked</param>
+    /// <returns><see cref="SecretsManagerOptions"/></returns>
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs:20]:
```suggestion
    /// <returns>Reference to the same <see cref="SecretsManagerOptions"/> to allow further configuration.</returns>
```

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+            return HealthCheckResult.Healthy();
+        }
+        catch (Exception ex)
+        {
+            return new HealthCheckResult(context.Registration.FailureStatus, exception: ex);
+        }
+    }
+
+    private IAmazonSecretsManager CreateSecretsManagerClient()
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs:34]: The same suggestion here as for #1128.

File: AspNetCore.Diagnostics.HealthChecks.sln (hunk 2 above)
Comment (sungam3r) [inline AspNetCore.Diagnostics.HealthChecks.sln:3]:
```suggestion
# Visual Studio Version 17
```

File: AspNetCore.Diagnostics.HealthChecks.sln [hunk 16]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio Version 17
-VisualStudioVersion = 17.0.31903.59
+VisualStudioVersion = 16.0.31829.152
```
Comment (sungam3r) [inline AspNetCore.Diagnostics.HealthChecks.sln:4]:
```suggestion
VisualStudioVersion = 17.0.31903.59
```

File: src/HealthChecks.Aws.SecretsManager/DependencyInjection/SecretsManagerHealthCheckBuilderExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    /// <param name="failureStatus">
+    /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+    /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+    /// </param>
+    /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+    /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
+    /// <returns>The specified <paramref name="builder"/>.</returns>
+    public static IHealthChecksBuilder AddSecretsManager(
+        this IHealthChecksBuilder builder,
+        Action<SecretsManagerOptions>? setup,
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/DependencyInjection/SecretsManagerHealthCheckBuilderExtensions.cs:28]: I doubt `SecretsManagerHealthCheck` is useful without specifying any configuration delegate.

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+public class SecretsManagerOptions
+{
+    public AWSCredentials? Credentials { get; set; }
+
+    public RegionEndpoint? RegionEndpoint { get; set; }
+
+    internal HashSet<string> _secrets = new HashSet<string>();
+
+    internal IEnumerable<string> Secrets => _secrets;
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs:14]:
```suggestion
    internal HashSet<string> Secrets { get; } = new HashSet<string>();
```

**Comment by sungam3r** [comment]
I added some minor fixes here as well as in #1128.

File: test/HealthChecks.Aws.SecretsManager.Tests/DependencyInjection/RegistrationTests.cs (hunk 13 above)
Comment (zodraz) [reply test/HealthChecks.Aws.SecretsManager.Tests/DependencyInjection/RegistrationTests.cs:9]: Fixed

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs (hunk 14 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs:20]: Fixed

File: AspNetCore.Diagnostics.HealthChecks.sln (hunk 16 above)
Comment (zodraz) [reply AspNetCore.Diagnostics.HealthChecks.sln]: Fixed

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs (hunk 18 above)
Comment (zodraz) [reply src/HealthChecks.Aws.SecretsManager/SecretsManagerOptions.cs:14]: Fixed

File: test/HealthChecks.Aws.SecretsManager.Tests/DependencyInjection/RegistrationTests.cs [hunk 19]
```diff
@@ -0,0 +1,123 @@
+using Amazon;
+using Amazon.Runtime;
+
+namespace HealthChecks.Aws.SecretsManager.Tests.DependencyInjection;
```
Comment (sungam3r) [inline test/HealthChecks.Aws.SecretsManager.Tests/DependencyInjection/RegistrationTests.cs:4]:
```suggestion
namespace HealthChecks.Aws.SecretsManager.Tests.DependencyInjection;

```

File: src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+    private async Task CheckSecretAsync(IAmazonSecretsManager client, string secretName, CancellationToken cancellationToken)
+    {
+        var request = new GetSecretValueRequest
+        {
+            SecretId = secretName,
+            VersionStage = "AWSCURRENT" // VersionStage defaults to AWSCURRENT if unspecified.
+        };
+
+        //Check the existence of the secret. If it does not throw it is a valid one (binary or not)
```
Comment (sungam3r) [inline src/HealthChecks.Aws.SecretsManager/SecretsManagerHealthCheck.cs:55]:
```suggestion
        // Check the existence of the secret. If it does not throw it is a valid one (binary or not)
```

File: src/HealthChecks.Aws.SecretsManager/DependencyInjection/SecretsManagerHealthCheckBuilderExtensions.cs (hunk 17 above)
Comment (sungam3r) [reply src/HealthChecks.Aws.SecretsManager/DependencyInjection/SecretsManagerHealthCheckBuilderExtensions.cs:28]: I'm fine either way.

**Review by sungam3r — APPROVED** [review]
Thanks.

**Comment by zodraz** [comment]
Thanks to you @sungam3r. I realized I did not update the main Readmemd file. Was that necessary?

**Comment by sungam3r** [comment]
I forgot about that. You may do it in new PR.

**Comment by sungam3r** [comment]
#1209

**Pull request merged**